### ISSUE TITLE
[RFC] Evaluation-calibrated adaptive KL for online policy updates

### ISSUE THREAD

**Issue by hkxxxxx** [Human-Agent-Society/reef#466]
### Primary area

Training and runtime

### Summary and decision to make

Add an opt-in, bounded adaptive-KL controller whose inner loop stabilizes online policy updates and whose outer loop is calibrated by held-out evaluation. The smallest decision is whether Reef should expose a controller/evaluator contract that can adjust KL pressure or clipping within safe bounds, while protected-task regressions, invalid evaluations, cost, and rollback remain release-gate decisions rather than hidden training signals.

### Motivation and supporting results

The current Reef recipes expose fixed stability controls such as KL coefficients, clipping, or loss weights, but there is no shared target-KL, adaptive-KL, or PID/dual-update controller. Fixed values are easy to reproduce, yet they are brittle across model scale, batch composition, reward variance, and multi-turn tool workloads. Too little constraint can cause destructive policy drift; too much constraint can make learning stall.

This proposal is grounded in established constrained-update ideas: TRPO formulates a trust-region constraint (https://arxiv.org/abs/1502.05477), PPO uses clipped policy updates (https://arxiv.org/abs/1707.06347), and KL-regularized language-model alignment uses a reference-policy penalty (https://arxiv.org/abs/1909.08593; https://arxiv.org/abs/2203.02155). The Reef-specific addition is to connect that optimization controller to the evaluation and protected-regression workflow in #355 and #356, so a low training KL is not mistaken for behavioral improvement.

### Goals and non-goals

- Provide a pluggable controller that keeps observed policy divergence and clipping within bounded, configurable ranges.
- Separate the fast training inner loop from a slower held-out evaluation outer loop.
- Use complete, valid evaluation results, protected-task vetoes, and cost/latency constraints when deciding whether to publish, pause, or roll back.
- Persist controller state and expose enough telemetry to reproduce and diagnose coefficient changes.
- Keep fixed-KL and fixed-clipping behavior as the default compatibility path.

Non-goals: declaring one universal target KL, replacing #355/#356 evaluation gates, using evaluator scores as an unbounded per-token reward, or claiming that KL alone measures agent quality.

### Proposal

Introduce an opt-in AdaptiveKLController with two explicit loops.

Inner training loop:

~~~text
observed KL, clip fraction, entropy, finite-loss status
                 -> bounded controller update
KL coefficient / clip range / optional learning-rate multiplier
~~~

A minimal bounded dual update is:

~~~text
lambda_next = clip(lambda * exp(eta * (observed_kl / target_kl - 1)),
                   lambda_min, lambda_max)
~~~

The controller should use an EMA or a clearly documented window for observed KL, update only after a successful finite training step, and support cooldown, max step size, coefficient bounds, and spike detection. A PID implementation may be added behind the same interface, but the first implementation should not require PID tuning. Each loss family may own an independent controller when its reference policy and divergence semantics differ; sharing state is opt-in only.

Outer evaluation loop:

~~~text
held-out score + confidence/validity + protected regressions + cost/latency
                              -> publish, pause, rollback, or recalibrate
~~~

Every calibration decision must identify the candidate, reference policy, evaluator/scorer versions, task split, and controller state. A complete evaluation with improvement and no protected regression may cautiously relax the constraint or increase the target within configured bounds; a regression, excessive cost, or unstable variance may tighten it or pause publication. An invalid or incomplete evaluation must not update the controller. Evaluation remains a gate and audit record, not an uncontrolled reward signal. The #356 confidence-aware paired selection proposal can supply statistical evidence, while #355 can supply historical and cross-release measurements.

### Public interfaces and configuration

Add an AdaptiveKLController interface with explicit state and update inputs, for example observed_kl, target_kl, clip_fraction, entropy, finite_loss, step, and reference_policy_id. The output should include the next KL coefficient, clip setting, optional learning-rate multiplier, update reason, and controller version.

Configuration should include target_kl, eta, bounds, smoothing window, cooldown, max update ratio, spike threshold, reference-policy kind, update cadence, and failure behavior. Add an EvaluationCalibrationPolicy interface that accepts valid held-out metrics, protected-regression results, cost/latency, and confidence metadata, then returns publish/pause/rollback/recalibrate plus a bounded target update. Persist both controller and calibration state in checkpoints with schema versions. Record whether each change came from the inner loop or outer loop.

### Compatibility and migration

The default remains the current fixed coefficient, clipping, and learning-rate behavior. Existing configuration names and checkpoints remain readable; enabling the controller is an explicit recipe-level flag. If a checkpoint has no controller state, initialize from configuration and record that initialization. If a controller state is incompatible with the reference policy, tokenizer, optimizer, or loss family, fail closed to the fixed path or stop according to the configured policy.

Rollout should start in telemetry-only mode, then shadow mode, then opt-in control for one recipe. A rollback consists of disabling the controller and serving the last accepted candidate; no training sample or existing scalar reward format needs migration.

### Security, privacy, and trust boundaries

Evaluator outputs, tool feedback, and user-provided task data are untrusted inputs. The controller must consume typed, bounded metrics rather than arbitrary text or executable evaluator instructions. Only authorized evaluator identities and versioned results may affect outer-loop calibration. Protected-task definitions and rollback permissions must remain explicit, and a malformed, spoofed, stale, or partial evaluation must not relax safeguards. Do not log raw prompts, tool results, credentials, or private evaluator data merely to explain a coefficient update.

### Operations and observability

Record target KL, observed KL, smoothed KL, KL coefficient, clip range, clip fraction, entropy, learning-rate multiplier, controller state/version, reference-policy identity, update reason, and spike/cooldown events. For the outer loop also record evaluator validity, confidence interval or uncertainty, held-out score delta, protected regressions, cost/latency delta, decision, and rollback reason. Add dashboards for divergence spikes, coefficient saturation, repeated pauses, evaluation invalid-rate, and disagreement between low KL and behavioral regression. Keep bounded histories suitable for checkpoint and postmortem inspection.

### Testing and documentation

- Unit-test the bounded dual update, smoothing, bounds, cooldown, spike detection, finite-loss guard, and deterministic state restoration.
- Test reference-policy identity and schema/version mismatches.
- Add controller tests showing that high KL increases constraint pressure and low KL relaxes it only within bounds.
- Add outer-loop fixtures for improvement, protected regression, evaluator outage, incomplete runs, noisy scores, excessive cost, and rollback.
- Verify invalid evaluations never change controller or calibration state.
- Run an offline replay or shadow-mode comparison against fixed-KL baselines and report stability, held-out quality, protected regressions, token cost, and latency.
- Document the research rationale, configuration, telemetry, and safe rollout example.

### Alternatives considered

Keep all coefficients fixed: maximally reproducible and simplest, but requires manual retuning across recipes and workloads.

Use only a training-side adaptive KL rule: can stabilize optimization, but may select a policy with low divergence and worse behavior.

Use evaluation score directly as a training reward: couples slow/noisy release evaluation to the inner loop and invites overfitting or reward hacking.

Use a full PID controller immediately: potentially responsive, but adds tuning and failure modes before Reef has a common telemetry and evaluation contract. A bounded dual update plus an explicit outer calibration policy is a smaller first step.

### Risks and unresolved questions

Observed KL depends on the reference policy, masking, token weighting, and aggregation; the RFC must define these semantics per loss family. A controller can oscillate, saturate at its bounds, or hide a bad reward model behind apparently healthy divergence. Held-out evaluation is expensive and can be noisy or distribution-shifted. Multiple controllers can also interact through a shared optimizer.

Open questions: whether the first implementation should control only KL coefficient or also clip range; whether target calibration should use confidence intervals, sequential tests, or a fixed decision budget; which metrics are mandatory for each recipe; and how TTTD group consistency and multi-turn reward events should constrain controller updates. These decisions should be resolved with maintainers of #355, #356, and the affected training recipes.

### Implementation plan and ownership

1. Specify divergence/reference-policy semantics and the typed controller state.
2. Implement the bounded inner-loop controller in telemetry-only mode with state persistence and unit tests.
3. Add the outer EvaluationCalibrationPolicy adapter using #355/#356 result contracts; keep invalid results non-updating.
4. Run shadow/offline comparisons against fixed settings, measuring KL stability, held-out quality, protected regressions, cost, and latency.
5. Enable the controller for one recipe behind an opt-in flag, with bounds, pause, and rollback; expand only after maintainer review.

Training/runtime maintainers review the controller and checkpoint changes. Evaluation maintainers review calibration, protected-task logic, and statistical decisions. Recipe owners review reference-policy and loss-family integration.

### Submission checks

- [x] I searched existing issues, pull requests, and RFCs for this decision.
- [x] I understand this issue is the RFC and that discussion is not acceptance.
- [x] I will keep this issue updated with material design changes and implementation links.

**Comment by BobbyZhouZijian** [comment]
Hi @hkxxxxx thanks for the PR! The adaptiveKL proposal is quite solid. Instead of an "opt-in" AdaptiveKLController, maybe this could be the default for any on-policy loss, used for SAO, OPD, etc.. After all, vanilla policy loss is almost not usable in practice. The outer evaluation loop could be a separate issue. Do you want to take it up and work on it?

**Comment by hkxxxxx** [comment]
Hi @BobbyZhouZijian, thanks for the feedback — I’d be happy to take this forward.

To clarify how this refines my original RFC: the original proposal combined a bounded inner-loop adaptive-KL controller with an evaluation-calibrated outer loop. Based on your suggestion, I propose splitting these into two linked work items rather than dropping the evaluation component.

To be explicit about the scope: I do want Reef to eventually support adaptive divergence control for OPD and SAO, but I do not propose applying the same verl-style scalar-β controller to both in the first PR.

For #466, I suggest making the controller capability available at the framework level for on-policy recipes, while keeping the divergence semantics loss-specific. In this sense, “default” would mean that a standard controller interface is available by default, not that PPO KL, OPD teacher divergence, and SAO policy-lag divergence share one coefficient or one controller blindly.

For the first end-to-end implementation, I propose targeting one clearly defined PPO/RLHF-style reference-policy path:

observed reference KL → bounded update of β → reward shaping for the next batch

Fixed-KL behavior would remain as the baseline and compatibility path. I would defer clip-range control until the basic controller semantics are validated.

For OPD, a follow-up should evaluate an OPD-specific entropy-aware or token-level teacher-student divergence controller, rather than directly reusing reward-side β shaping. Existing work such as EOPD and AED already studies entropy-conditioned or token-level adaptive divergence in OPD.

For SAO, we should first specify whether the constraint is against a frozen reference policy or the rollout policy. The latter would overlap with SAO’s existing policy-lag importance sampling and token-level clipping mechanism, so it should not be treated as the same reference-policy KL problem.

Therefore, OPD and SAO should be integrated through separate adapters and validated through separate experiments, rather than being forced to share one default KL coefficient.

The evaluation side — held-out evaluation, protected-task regressions, cost/latency checks, pause, and rollback — can be moved to a separate issue linked to #355 and #356. This would be a scope and ownership split, not a change to the original motivation: evaluation should remain a release and safety gate, rather than becoming a hidden training reward.

I have also opened the corresponding training-side discussion in Slime #2387. Would this split and the phased implementation plan make sense?

**Comment by BobbyZhouZijian** [comment]
@hkxxxxx thanks for the detailed explanation! #533 has laid some preliminary groundwork by adding some key components of the distillation family into reef. Maybe @MikeZheng777 can add some comments here.

**Comment by hkxxxxx** [comment]
Thanks Bobby for pointing out #533. We agree that it provides useful distillation groundwork, but it does not cover the PPO/RLHF-style reference-policy adaptive KL discussed here.

We are moving forward with a focused, opt-in implementation of reward-side adaptive KL:

`shaped_reward = task_reward - beta * reference_KL`

The first phase will keep OPD/teacher divergence, SAO `ppo_kl`, GRPO reward-side KL, and loss-side `kl_loss_coef` separate. We will validate the controller in telemetry/shadow mode, keep the default behavior unchanged, and open a focused PR with tests and the exact Slime runtime details. The distillation infrastructure from #533 can remain the foundation for future OPD-specific work.
